servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java

```diff
@@ -16,72 +16,95 @@
 package tachyon.worker.block.evictor;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collections;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 
 import tachyon.Pair;
 import tachyon.worker.block.BlockMetadataManager;
 import tachyon.worker.block.BlockStoreLocation;
 import tachyon.worker.block.meta.BlockMeta;
+import tachyon.worker.block.meta.StorageDir;
 
 public class EvictorUtils {
   /**
-   * Checks whether the plan is legal for an evictor with cascading eviction feature like
+   * Checks whether the plan is valid for an evictor with cascading eviction feature like
    * {@link LRUEvictor}. The plan is legal when the requested space can be satisfied, and there is
    * enough space in lower tier to move in blocks from upper tier.
    *
    * @param bytesToBeAvailable requested bytes to be available after eviction
-   * @param plan the eviction plan
+   * @param plan the eviction plan, should not be empty
    * @param metaManager the meta data manager
    * @return true if the above requirements are satisfied, otherwise false.
    */
-  public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan plan,
+  public static boolean validCascadingPlan(long bytesToBeAvailable, EvictionPlan plan,
       BlockMetadataManager metaManager) throws IOException {
     // reassure the plan is feasible: enough free space to satisfy bytesToBeAvailable, and enough
     // space in lower tier to move blocks in upper tier there
-    Map<Integer, Long> bytesToBeAvailableInTier =
-        new HashMap<Integer, Long>(metaManager.getTiers().size());
-    List<Integer> tierAliases = new ArrayList<Integer>();
 
-    List<Long> blockIds = new ArrayList<Long>(plan.toEvict().size() + plan.toMove().size());
-    blockIds.addAll(plan.toEvict());
-    for (Pair<Long, BlockStoreLocation> move : plan.toMove()) {
-      blockIds.add(move.getFirst());
+    // Map from dir to a pair of bytes to be available in this dir and bytes to move into this dir
+    // after the plan taking action
+    Map<StorageDir, Pair<Long, Long>> spaceInfoInDir = new HashMap<StorageDir, Pair<Long, Long>>();
+
+    for (long blockId : plan.toEvict()) {
+      BlockMeta block = metaManager.getBlockMeta(blockId);
+      StorageDir dir = block.getParentDir();
+      if (spaceInfoInDir.containsKey(dir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(dir);
+        spaceInfo.setFirst(spaceInfo.getFirst() + block.getBlockSize());
+        spaceInfoInDir.put(dir, spaceInfo);
```

Comment:
I don't think line 54 is necessary as `spaceInfo` is the reference to the `Pair`, could you doublecheck?